archive/demos/shiny-drum-machine-ui.js

```diff
@@ -0,0 +1,272 @@
+class Slider {
+    constructor(element, opts = {}) {
+        this.element = element;
+        this.onchange = () => {};
+
+        element.addEventListener('input', () => {
+            this.onchange(this.value);
+        });
+
+        if(opts && opts.doubleClickValue) {
+            element.addEventListener('dblclick', () => {
+                this.value = opts.doubleClickValue;
+                this.onchange(this.value);
+            });
+        }
+    }
+
+    get value() {
+        return new Number(this.element.value) / 100;
+    }
+
+    set value(value) {
+        this.element.value = value * 100;
+    }
+}
+
+class EffectSlider extends Slider {
+    constructor(container=document) {
+        super(container.getElementById('effect_thumb'));
+    }
+}
+
+class SwingSlider extends Slider {
+    constructor(container=document) {
+        super(container.getElementById('swing_thumb'));
+    }
+}
+
+class PitchSliders {
+    constructor(container=document) {
+        this.sliders = {};
+        this.onPitchChange = (instrument, pitch) => {};
+        for(const el of container.querySelectorAll('[data-instrument][data-pitch]')) {
+            this.sliders[el.dataset.instrument] = new Slider(el, {doubleClickValue: 0.5});
+            this.sliders[el.dataset.instrument].onchange = (value) => {
+                this.onPitchChange(el.dataset.instrument, value);
+            };
+        }
+    }
+
+    setPitch(instrument, pitch) {
+        this.sliders[instrument].value = pitch;
+    }
+}
+
+class Button {
+    constructor(element, onclick = () => {}) {
+        this.element = element;
+        this.onclick = onclick;
+        element.addEventListener('click', (event) => this.onclick(event));
+    }
+
+    get state() {
+        return this.element.dataset.state;
+    }
+
+    set state(value) {
+        this.element.dataset.state = value;
+    }
+}
+
+class PlayButton extends Button {
+    constructor() {
+        super(document.getElementById('play'));
+    }
+}
+
+class ResetButton extends Button {
+    constructor() {
+        super(document.getElementById('reset'));
+    }
+}
+
+class Playheads {
+    constructor() {
+        this.current = 0;
+    }
+
+    drawPlayhead(index) {
+        // TODO: Refactor, replace with data attributes.
```

Comment:
If we add a TODO, we might want to tie this with an issue on the issue tracker. Otherwise we won't have the visibility on these "TODOs in the code".

Reply:
Added reference to #226.